syntax = "proto3";

package kmeans;

service KMeansCluster {
  rpc SendDataToMapper(MapperRequest) returns (MapperResponse);

  rpc ReceiveUpdatedCentroids(ReducerResponse) returns (Acknowledgement);

  rpc ProcessDataForReducer(ReducerRequest) returns (ReducerResponse);

  rpc send_intermediate_values_to_reducer(IntermediateRequest) returns (IntermediateResponse);
}

message MapperRequest {
  int32 mapper_id = 1;
  repeated Centroid centroids = 2;
  int32 range_start = 3;
  int32 range_end = 4;
  int32 num_red = 5;
}

message MapperResponse {
  int32 mapper_id = 1;
  string status = 2; // SUCCESS or FAILURE
}

message ReducerRequest {
  int32 reducer_id = 1;
  int32 num_mappers = 2;
}

message ReducerResponse {
  int32 reducer_id = 1;
  repeated Centroid new_centroids = 2;
  string status = 3; // SUCCESS or FAILURE
}

message IntermediateRequest {
  int32 reducer_id = 1;
  int32 num_mappers = 2;
}

message IntermediateResponse {
  int32 reducer_id = 1;
  repeated string data = 2;
}

// Acknowledgement to reducer response
message Acknowledgement {
  string message = 1; 
}

message Centroid {
  repeated double coordinates = 1;
}
